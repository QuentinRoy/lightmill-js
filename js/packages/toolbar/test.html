<html>
    
    <head>
        <title>Toolbar test</title>
        <script src="../../libs/require.js"></script>
        <style>
            .full-parent {
                position:absolute;
                width:100%;
                height:100%;
                margin:0;
                padding:0;
                top:0;
                left:0;
            }
            #message {
                text-align:center
            }
        </style>
        <script>
            requirejs.config({
                baseUrl: "../..",
                enforceDefine: true,
                paths: {
                    jquery: 'libs/jquery-2.0.0',
                    text: 'libs/text',
                    handlebars: 'libs/handlebars',
                },
                map: {
                    '*': {
                        'css': 'packages/require-css/css' // or whatever the path to require-css is
                    }
                },
                packages: [{
                        name: 'toolbar',
                        location: 'packages/toolbar/',
                        main: 'toolbar'
                    }
                ],

                shim: {
                    handlebars: {
                        exports: 'Handlebars'
                    },
                }
            });


            require(['toolbar', 'jquery'], function(ToolBar, $, _) {
                'use strict';


                var messageDiv = $('#message'),
                    toolbar, buttons = ['button1', 'button2 et du texte', 'button3', 'button4'],
                    callbacks = {};

                $.each(buttons, function(index, buttonName) {
                    callbacks[buttonName] = function() {
                        messageDiv.html(messageDiv.html() + buttonName + '<br/>');
                    }
                });

                toolbar = new ToolBar($('#main-div'), callbacks, {
                    buttonWidth: 100,
                    spread:true
                });
            });
        </script>
    </head>
    
    <body>
        <div id="message" class="full-parent"></div>
        <div id="main-div" class="full-parent"></div>
    </body>

</html>