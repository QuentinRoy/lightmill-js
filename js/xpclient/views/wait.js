/*jslint nomen: true, browser:true*/
/*global define */

define(['spin', 'jquery'], function (Spinner, $) {
    "use strict";

    function WaitView(parentView) {
        this._parentView = parentView;
        this._spinView = $('<div class="full-parent"></div>');
        this._spinner = new Spinner(this._opts);
    }

    WaitView.prototype = {

        _opts: {
            lines: 11, // The number of lines to draw
            length: 20, // The length of each line
            width: 10, // The line thickness
            radius: 30, // The radius of the inner circle
            corners: 1, // Corner roundness (0..1)
            rotate: 0, // The rotation offset
            direction: 1, // 1: clockwise, -1: counterclockwise
            color: '#5C5C5C', // #rgb or #rrggbb or array of colors
            speed: 1, // Rounds per second
            trail: 60, // Afterglow percentage
            shadow: false, // Whether to render a shadow
            hwaccel: false, // Whether to use hardware acceleration
            className: 'spinner', // The CSS class to assign to the spinner
            zIndex: 2e9, // The z-index (defaults to 2000000000)
            top: 'auto', // Top position relative to parent in px
            left: 'auto' // Left position relative to parent in px
        },

        startWaiting: function () {
            this._spinView.appendTo(this._parentView);
            this._spinner.spin(this._spinView[0]);
        },
        stopWaiting: function () {
            this._spinner.stop();
            this._spinView.remove();
        }
    };

    return WaitView;

});