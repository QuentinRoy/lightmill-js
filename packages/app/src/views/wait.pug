.lightmill-base-app.lightmill-base-app-wait
  .lightmill-puck-wrapper
    .lightmill-puck.lightmill-puck-1
    .lightmill-puck.lightmill-puck-2
