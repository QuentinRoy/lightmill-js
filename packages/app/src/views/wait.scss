@import './common';

$puck-wrapper-size: 60px;
$puck-radius: 5px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .#{$class-prefix}puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    animation: $animation-name $rotation-duration infinite linear;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .#{$class-prefix}puck {
      position: absolute;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }

    .#{$class-prefix}puck-1 {
      top: 0;
      left: 0;
    }

    .#{$class-prefix}puck-2 {
      right: 0;
      bottom: 0;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
