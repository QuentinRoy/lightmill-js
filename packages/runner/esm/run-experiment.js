import { RunInterface } from '@lightmill/connection';
import runTrials from './run-trials';
import LocalValueStorage from './local-value-storage';
import { throwWithHeader } from './utils';

/**
 * Run an experiment.
 *
 * @param {Object} app The application of the experiment.
 * @param {function(): Promise} [app.start] Initialize the experiment.
 * @param {function(): Promise} [app.initRun] Initialize the run.
 * @param {function(): Promise} [app.initBlock] Initialize a block.
 * @param {function(): Promise} app.runTrial Run a trial.
 * @param {function(): Promise} [app.end] Called when the experiment ends.
 * @param {function(): Promise} [app.crash] Called if an error is thrown during
 * the run.
 * @param {Object} config Configuration.
 * @param {string} config.experimentId The id of the experiment. This is
 * required is config.connection is not provided.
 * @param {string} [config.runId] The id of a run to connect to.
 * @param {string} [config.serverAddress] The address of the xp server.
 * @param {string} [config.experimentDesignAddr] The path toward a touchstone
 * experiment design xml file.
 * @param {string} [config.queueSize] The maximum number of pending trial result
 * posts before starting a new trial.
 * @param {Object} [config.runStorage] Used to store the running run.
 * @param {Object} [config.connection] The connection to the server. You do not
 * usually need to provide this.
 * @returns {Promise} A promise resolved once the experiment is done.
 */
export default async (
  app,
  {
    experimentId,
    runId,
    serverAddress,
    experimentDesignAddr,
    queueSize = 1,
    // Test arguments.
    runStorage = new LocalValueStorage(`${experimentId}/running-run-id`),
    connection: potentialConnection = RunInterface.create(
      serverAddress || `http://${window.location.hostname}:5000`,
      experimentId,
      runId || runStorage.get(),
      experimentDesignAddr
    )
  }
) => {
  let run_; // eslint-disable-line no-underscore-dangle
  try {
    // Returns a promise that initializes the connection and registers the run.
    const initConnection = async () => {
      const connection = await potentialConnection;
      const run = await connection.getRun();
      runStorage.set(run.id);
      return { connection, run };
    };
    // Connect to the run and start the app.
    const [{ connection, run }] = await Promise.all([
      initConnection().catch(
        throwWithHeader('Could not connect to the experiment')
      ),
      // Start the experiment app.
      Promise.resolve(app.start && app.start()).catch(
        throwWithHeader("Could not start the experiment's app")
      )
    ]);
    // Export the run out of the context for the catch clause.
    run_ = run;
    // Ask the app to init the run.
    await Promise.resolve(app.initRun && app.initRun(run)).catch(
      throwWithHeader('Could not init the run task')
    );
    // Run the trials.
    await runTrials(connection, app, queueSize);
    // Clear the local storage so that next time the page is loaded, a new run
    // will be requested.
    runStorage.remove();
    // Disconnect from the server.
    await connection.disconnect();
    // Notify the app that the experiment is finished.
    if (app.end) await app.end();
  } catch (e) {
    if (app.crash) app.crash(e.message, e, run_);
    throw e;
  }
};
