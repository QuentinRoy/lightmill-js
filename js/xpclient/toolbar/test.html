<html>
    
    <head>
        <title>Toolbar test</title>
        <script src="../../libs/require.js"></script>
        <style>
            .full-parent {
                position:absolute;
                width:100%;
                height:100%;
                margin:0;
                padding:0;
                top:0;
                left:0;
            }
        </style>
        <script>
            requirejs.config({
                baseUrl: "../..",
                enforceDefine: true,
                paths: {
                    'jquery': 'libs/jquery-2.0.0',
                    'text': 'libs/text',
                    'jstools': 'packages/jstools/src',
                    'underscore': 'libs/underscore',
                    'classy': 'libs/classy-modified',
                },
                map: {
                    '*': {
                        'css': 'packages/require-css/css' // or whatever the path to require-css is
                    }
                },

                shim: {
                    underscore: {
                        exports: '_'
                    },
                }
            });


            require(['xpclient/toolbar/toolbar', 'jquery'], function(ToolBar, $) {
                'use strict';
                toolbar = new ToolBar($('#main-div'), {
                    button1: function() {
                        alert('button1');
                    },
                    button2: function() {
                        alert('button2');
                    },
                    button3: function() {
                        alert('button3');
                    },
                })
            });
        </script>
    </head>
    
    <body>
        <div id="main-div" class="full-parent"></div>
    </body>

</html>