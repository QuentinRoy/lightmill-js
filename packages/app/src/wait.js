/* eslint-disable no-param-reassign */

import waitTemplate from './views/wait.pug';
import './views/wait.scss';

export default (node, message) => {
  node.innerHTML = waitTemplate({ message });

  return {
    remove() {
      node.innerHTML = '';
    }
  };
};
