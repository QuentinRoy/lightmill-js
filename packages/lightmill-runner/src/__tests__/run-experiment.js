import test from 'ava';
import { spy } from 'sinon';
import wait from 'wait-then';
import range from 'array-range';
import deferred from 'promise.defer';
import { runTrials, runExperiment } from '../run-experiment';

const makeTrialList = (blockCount = 2, trialCountPerBlock = 3) =>
  range(blockCount).reduce((acc, blockNumber) => {
    const block = { number: blockNumber };
    return [
      ...acc,
      ...range(trialCountPerBlock).map(number => ({ number, block }))
    ];
  }, []);

// Create a group of delayed spies, each spies create from this group is given a type
// and all spy calls can be monitored (in order) with may `all` spy.
// Also, all spies return a promise that resolves asynchronously. The resolution of the promise
// can be monitored as well.
const createDelayedSpiesGroup = () => {
  const all = spy();
  const createSpy = (
    type,
    f,
    { delay = 0, done = delay || delay === 0 } = {}
  ) =>
    spy((...args) => {
      all(type, args);
      let res;
      if (!delay && delay !== 0) {
        res = typeof f === 'function' ? f(...args) : f;
      } else {
        res = wait(delay).then(
          () => (typeof f === 'function' ? f(...args) : f)
        );
      }
      if (done) {
        res = res.then(
          res_ => {
            all(`* ${type}`, res_);
            return res_;
          },
          err => {
            all(`X ${type}`, err);
            throw err;
          }
        );
      }
      return res;
    });
  return { all, spy: createSpy };
};

test('`runTrials` does not wait for post results before starting new trials', async t => {
  const postDefers = range(3).map(() => deferred());
  const postPromises = postDefers.map(def => def.promise);
  const trials = makeTrialList(1, 3);
  const connection = {
    postResults: spy(() => Promise.resolve(postPromises.shift())),
    endTrial: spy(() => Promise.resolve(trials.shift())),
    getCurrentTrial: spy(() => Promise.resolve(trials.shift())),
    flush: spy(() => Promise.resolve())
  };
  const app = { runTrial: spy(() => Promise.resolve({})) };
  runTrials(connection, app);
  await wait();
  t.deepEqual(
    app.runTrial.args,
    [
      [{ number: 0, block: { number: 0 } }],
      [{ number: 1, block: { number: 0 } }],
      [{ number: 2, block: { number: 0 } }]
    ],
    '`runTrial` have been called as expected'
  );
});

test('`runTrials` throws if a post fails during a trial', async t => {
  const postDefers = range(4).map(() => deferred());
  const postPromises = postDefers.map(def => def.promise);
  const trialDefers = range(4).map(() => deferred());
  const trialPromises = trialDefers.map(def => def.promise);
  const trials = makeTrialList(1, 4);
  const connection = {
    postResults: spy(() => postPromises.shift()),
    endTrial: spy(() => Promise.resolve(trials.shift())),
    getCurrentTrial: spy(() => Promise.resolve(trials.shift())),
    flush: spy(() => Promise.resolve())
  };
  const app = { runTrial: spy(() => trialPromises.shift()) };
  const prom = runTrials(connection, app);
  trialDefers[0].resolve({ res: 1 });
  trialDefers[1].resolve({ res: 2 });
  postDefers[0].resolve();
  postDefers[1].reject(new Error('test error'));
  await t.throws(
    prom,
    'Could not post trial results: test error',
    '`runTrials` threw'
  );
  t.deepEqual(
    connection.postResults.args,
    [[{ res: 1 }], [{ res: 2 }]],
    '`postResults` have been called as expected'
  );
  t.is(
    connection.endTrial.callCount,
    2,
    '`endTrial` have been called as expected'
  );
  t.deepEqual(
    app.runTrial.args,
    [
      [{ number: 0, block: { number: 0 } }],
      [{ number: 1, block: { number: 0 } }],
      [{ number: 2, block: { number: 0 } }]
    ],
    '`runTrial` have been called as expected'
  );
});

test('`runTrials` throws if a post fails even if all trials are done', async t => {
  const postDefers = range(4).map(() => deferred());
  const postPromises = postDefers.map(def => def.promise);
  const trialPromises = range(4).map(i => Promise.resolve({ res: i + 1 }));
  const trials = makeTrialList(1, 4);
  const connection = {
    postResults: spy(() => postPromises.shift()),
    endTrial: spy(() => Promise.resolve(trials.shift())),
    getCurrentTrial: spy(() => Promise.resolve(trials.shift())),
    flush: spy(
      () =>
        trialPromises.length > 0 ? Promise.resolve() : new Promise(() => {})
    )
  };
  const app = { runTrial: spy(() => trialPromises.shift()) };
  const prom = runTrials(connection, app);
  (async () => {
    await wait();
    postDefers[0].resolve();
    await wait();
    postDefers[1].resolve();
    await wait();
    postDefers[2].reject(new Error('test error'));
  })();
  await t.throws(
    prom,
    'Could not post trial results: test error',
    '`runTrials` threw'
  );
  t.deepEqual(
    connection.postResults.args,
    [[{ res: 1 }], [{ res: 2 }], [{ res: 3 }], [{ res: 4 }]],
    '`postResults` have been called as expected'
  );
  t.is(
    connection.endTrial.callCount,
    4,
    '`endTrial` have been called as expected'
  );
  t.deepEqual(
    app.runTrial.args,
    [
      [{ number: 0, block: { number: 0 } }],
      [{ number: 1, block: { number: 0 } }],
      [{ number: 2, block: { number: 0 } }],
      [{ number: 3, block: { number: 0 } }]
    ],
    '`runTrial` have been called as expected'
  );
});

test('`runExperiment` runs as expected if every app handlers are provided', async t => {
  const spyGroup = createDelayedSpiesGroup();
  // Connection fake implementation.
  const trials = makeTrialList(2, 2);
  const connection = {
    disconnect: spyGroup.spy('connection.disconnect'),
    postResults: spyGroup.spy('connection.postResults', null, { done: false }),
    endTrial: spyGroup.spy('connection.endTrial', () => trials.shift()),
    getCurrentTrial: spyGroup.spy('connection.getCurrentTrial', () =>
      trials.shift()
    ),
    flush: spyGroup.spy('connection.flush'),
    getRun: spyGroup.spy('connection.getRun', { id: 'runId' })
  };
  // App fake implementation.
  const app = {
    start: spyGroup.spy('app.start', null, { done: null }),
    initBlock: spyGroup.spy('app.initBlock'),
    runTrial: spyGroup.spy('app.runTrial', trial => ({
      result: `${trial.block.number}-${trial.number}`
    })),
    end: spyGroup.spy('app.end', null, { delay: null })
  };
  // Storage fake implementation
  const runStorage = {
    set: spyGroup.spy('storage.set', null, { delay: null }),
    remove: spyGroup.spy('storage.remove', null, { delay: null })
  };
  await runExperiment(app, {
    experimentId: 'xpId',
    runId: 'runId',
    connection,
    runStorage
  });
  t.snapshot(spyGroup.all.args);
});

test('`runExperiment` works properly even if only `app.runTrial` is provided', async t => {
  const spyGroup = createDelayedSpiesGroup();
  // Connection fake implementation.
  const trials = makeTrialList(2, 2);
  const connection = {
    disconnect: spyGroup.spy('connection.disconnect'),
    postResults: spyGroup.spy('connection.postResults', null, { done: false }),
    endTrial: spyGroup.spy('connection.endTrial', () => trials.shift()),
    getCurrentTrial: spyGroup.spy('connection.getCurrentTrial', () =>
      trials.shift()
    ),
    flush: spyGroup.spy('connection.flush'),
    getRun: spyGroup.spy('connection.getRun', { id: 'runId' })
  };
  // App fake implementation.
  const app = {
    runTrial: spyGroup.spy('app.runTrial', trial => ({
      result: `${trial.block.number}-${trial.number}`
    }))
  };
  // Storage fake implementation
  const runStorage = {
    set: spyGroup.spy('storage.set', null, { delay: null }),
    remove: spyGroup.spy('storage.remove', null, { delay: null })
  };
  await runExperiment(app, {
    experimentId: 'xpId',
    runId: 'runId',
    connection,
    runStorage
  });
  t.snapshot(spyGroup.all.args);
});

test('`runExperiment` calls `app.crash` then throws if a post goes wrong', async t => {
  const spyGroup = createDelayedSpiesGroup();
  const trials = makeTrialList(2, 2);
  // Deferred and promise lists for result posts.
  const postDefs = trials.map(() => deferred());
  const postProms = postDefs.map(def => def.promise);
  // Deferred and promise lists for trial results.
  const trialDefs = trials.map(() => deferred());
  const trialProms = trialDefs.map(def => def.promise);
  // Flush getters, defines the behavior of `connection.flush`.
  const flushGetters = [
    ...trials.map(() => () => Promise.resolve()),
    () => new Promise(() => {})
  ];
  // Connection fake implementation.
  const connection = {
    disconnect() {},
    postResults: spyGroup.spy('postResults', () => postProms.shift(), {
      delay: null,
      done: true
    }),
    endTrial: async () => trials.shift(),
    getCurrentTrial: async () => trials.shift(),
    flush: () => flushGetters.shift().call(),
    getRun: () => wait().then(() => ({ id: 'runId' }))
  };
  // App fake implementation.
  const app = {
    runTrial: spyGroup.spy('runTrial', () => trialProms.shift(), {
      delay: null,
      done: true
    }),
    crash: spyGroup.spy('crash')
  };
  // Run the experiment.
  const runP = t.throws(
    runExperiment(app, {
      experimentId: 'xpId',
      runId: 'runId',
      connection,
      runStorage: { set() {}, remove() {} }
    }),
    'Could not post trial results: reject test'
  );
  trialDefs[0].resolve();
  trialDefs[1].resolve();
  await wait(5);
  postDefs[0].resolve();
  postDefs[1].reject(new Error('reject test'));
  await runP;
  t.snapshot(spyGroup.all.args);
});

test('`runExperiment` calls `app.crash` then throws if a trial goes wrong', async t => {
  const spyGroup = createDelayedSpiesGroup();
  const trials = makeTrialList(2, 2);
  // Connection fake implementation.
  const connection = {
    disconnect() {},
    postResults: () => wait(),
    endTrial: async () => trials.shift(),
    getCurrentTrial: async () => trials.shift(),
    flush: () => wait(),
    async getRun() {
      await wait();
      return { id: 'runId' };
    }
  };
  // App fake implementation.
  const runTrialGetters = [
    () => Promise.resolve({ res: 1 }),
    () => Promise.resolve({ res: 2 }),
    () => Promise.reject(new Error('bad trial'))
  ];
  const app = {
    runTrial: spyGroup.spy('runTrial', () => runTrialGetters.shift().call()),
    crash: spyGroup.spy('crash')
  };
  // Run the experiment.
  await t.throws(
    runExperiment(app, {
      experimentId: 'xpId',
      runId: 'runId',
      connection,
      runStorage: { set() {}, remove() {} }
    }),
    'bad trial'
  );
  t.snapshot(spyGroup.all.args);
});

test('`runExperiment` calls `app.crash` then throws if a post goes wrong even if all trials are done', async t => {
  const spyGroup = createDelayedSpiesGroup();
  // Flat list of all trials.
  const trials = makeTrialList(2, 1);
  // Deferred and promise lists for result posts.
  const postDefs = trials.map(() => deferred());
  const posts = postDefs.map(def => def.promise);
  // Flush getters, defines the behavior of `connection.flush`.
  const flushGetters = [
    ...trials.map(() => () => wait()),
    () => new Promise(() => {})
  ];
  // Connection fake implementation.
  const connection = {
    disconnect() {},
    postResults: spyGroup.spy('postResults', () => posts.shift()),
    endTrial: async () => trials.shift(),
    getCurrentTrial: async () => trials.shift(),
    flush: () => flushGetters.shift().call(),
    getRun: () => wait().then(() => ({ id: 'runId' }))
  };
  // Resolved when all trials are done.
  const trialDone = deferred();
  // App fake implementation.
  const results = range(trials.length);
  const app = {
    runTrial: spyGroup.spy('runTrial', () => {
      const res = results.shift();
      if (results.length === 0) trialDone.resolve();
      return { res };
    }),
    crash: spyGroup.spy('crash')
  };
  // Run the experiment.
  const runP = t.throws(
    runExperiment(app, {
      experimentId: 'xpId',
      runId: 'runId',
      connection,
      queueSize: 5,
      runStorage: { set() {}, remove() {} }
    }),
    'Could not post trial results: reject test',
    '`runExperiment` threw.'
  );
  await trialDone.promise;
  // Resolve one posts.
  postDefs[0].resolve();
  postDefs[1].reject(new Error('reject test'));

  await runP;

  t.snapshot(spyGroup.all.args);
});

test.todo(
  '`runExperiment` calls `app.crash` then throws for any possible things that can go wrong.'
);

test.todo(
  '`runExperiment` creates a RunInterface using stored run id if none are provided'
);
