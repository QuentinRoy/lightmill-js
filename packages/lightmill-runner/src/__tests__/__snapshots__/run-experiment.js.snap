// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`runExperiment\` calls \`app.crash\` then throws if a post goes wrong 1`] = `
Array [
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* runTrial",
    undefined,
  ],
  Array [
    "postResults",
    Array [
      undefined,
    ],
  ],
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* runTrial",
    undefined,
  ],
  Array [
    "postResults",
    Array [
      undefined,
    ],
  ],
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* postResults",
    undefined,
  ],
  Array [
    "X postResults",
    [Error: reject test],
  ],
  Array [
    "crash",
    Array [
      "Could not post trial results: reject test",
      [Error: Could not post trial results: reject test],
      Object {
        "id": "runId",
      },
    ],
  ],
]
`;

exports[`\`runExperiment\` calls \`app.crash\` then throws if a post goes wrong even if all trials are done 1`] = `
Array [
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* runTrial",
    Object {
      "res": 0,
    },
  ],
  Array [
    "postResults",
    Array [
      Object {
        "res": 0,
      },
    ],
  ],
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* runTrial",
    Object {
      "res": 1,
    },
  ],
  Array [
    "postResults",
    Array [
      Object {
        "res": 1,
      },
    ],
  ],
  Array [
    "* postResults",
    undefined,
  ],
  Array [
    "X postResults",
    [Error: reject test],
  ],
  Array [
    "crash",
    Array [
      "Could not post trial results: reject test",
      [Error: Could not post trial results: reject test],
      Object {
        "id": "runId",
      },
    ],
  ],
]
`;

exports[`\`runExperiment\` calls \`app.crash\` then throws if a trial goes wrong 1`] = `
Array [
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* runTrial",
    Object {
      "res": 1,
    },
  ],
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* runTrial",
    Object {
      "res": 2,
    },
  ],
  Array [
    "runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "X runTrial",
    [Error: bad trial],
  ],
  Array [
    "crash",
    Array [
      "bad trial",
      [Error: bad trial],
      Object {
        "id": "runId",
      },
    ],
  ],
]
`;

exports[`\`runExperiment\` runs as expected 1`] = `
Array [
  Array [
    "app.start",
    Array [],
  ],
  Array [
    "connection.getRun",
    Array [],
  ],
  Array [
    "* connection.getRun",
    Object {
      "id": "runId",
    },
  ],
  Array [
    "storage.set",
    Array [
      "runId",
    ],
  ],
  Array [
    "app.initBlock",
    Array [
      Object {
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.initBlock",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "0-0",
    },
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "0-1",
    },
  ],
  Array [
    "app.initBlock",
    Array [
      Object {
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.initBlock",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "1-0",
    },
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "1-1",
    },
  ],
  Array [
    "storage.remove",
    Array [],
  ],
  Array [
    "connection.disconnect",
    Array [],
  ],
  Array [
    "* connection.disconnect",
    undefined,
  ],
  Array [
    "app.end",
    Array [],
  ],
]
`;

exports[`\`runExperiment\` runs as expected if every app handlers are provided 1`] = `
Array [
  Array [
    "app.start",
    Array [],
  ],
  Array [
    "connection.getRun",
    Array [],
  ],
  Array [
    "* connection.getRun",
    Object {
      "id": "runId",
    },
  ],
  Array [
    "storage.set",
    Array [
      "runId",
    ],
  ],
  Array [
    "connection.getCurrentTrial",
    Array [],
  ],
  Array [
    "* connection.getCurrentTrial",
    Object {
      "block": Object {
        "number": 0,
      },
      "number": 0,
    },
  ],
  Array [
    "app.initBlock",
    Array [
      Object {
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.initBlock",
    undefined,
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "0-0",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "0-0",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    Object {
      "block": Object {
        "number": 0,
      },
      "number": 1,
    },
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "0-1",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "0-1",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    Object {
      "block": Object {
        "number": 1,
      },
      "number": 0,
    },
  ],
  Array [
    "app.initBlock",
    Array [
      Object {
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.initBlock",
    undefined,
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "1-0",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "1-0",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    Object {
      "block": Object {
        "number": 1,
      },
      "number": 1,
    },
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "1-1",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "1-1",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    undefined,
  ],
  Array [
    "connection.flush",
    Array [],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "storage.remove",
    Array [],
  ],
  Array [
    "connection.disconnect",
    Array [],
  ],
  Array [
    "* connection.disconnect",
    undefined,
  ],
  Array [
    "app.end",
    Array [],
  ],
]
`;

exports[`\`runExperiment\` works properly even if only \`app.runTrial\` is provided 1`] = `
Array [
  Array [
    "connection.getRun",
    Array [],
  ],
  Array [
    "* connection.getRun",
    Object {
      "id": "runId",
    },
  ],
  Array [
    "storage.set",
    Array [
      "runId",
    ],
  ],
  Array [
    "connection.getCurrentTrial",
    Array [],
  ],
  Array [
    "* connection.getCurrentTrial",
    Object {
      "block": Object {
        "number": 0,
      },
      "number": 0,
    },
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "0-0",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "0-0",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    Object {
      "block": Object {
        "number": 0,
      },
      "number": 1,
    },
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 0,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "0-1",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "0-1",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    Object {
      "block": Object {
        "number": 1,
      },
      "number": 0,
    },
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 0,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "1-0",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "1-0",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    Object {
      "block": Object {
        "number": 1,
      },
      "number": 1,
    },
  ],
  Array [
    "connection.flush",
    Array [
      1,
    ],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "app.runTrial",
    Array [
      Object {
        "block": Object {
          "number": 1,
        },
        "number": 1,
      },
    ],
  ],
  Array [
    "* app.runTrial",
    Object {
      "result": "1-1",
    },
  ],
  Array [
    "connection.postResults",
    Array [
      Object {
        "result": "1-1",
      },
    ],
  ],
  Array [
    "connection.endTrial",
    Array [],
  ],
  Array [
    "* connection.endTrial",
    undefined,
  ],
  Array [
    "connection.flush",
    Array [],
  ],
  Array [
    "* connection.flush",
    undefined,
  ],
  Array [
    "storage.remove",
    Array [],
  ],
  Array [
    "connection.disconnect",
    Array [],
  ],
  Array [
    "* connection.disconnect",
    undefined,
  ],
]
`;
